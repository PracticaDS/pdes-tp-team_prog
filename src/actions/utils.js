export const BEGIN = 'BEGIN'
export const SUCCESS = 'SUCCESS'
export const ERROR = 'ERROR'
export const GET = 'GET'
export const POST = 'POST'
export const PUT = 'PUT'
export const DELETE = 'DELETE'

export const customFetch = (api, req) =>
  new Promise((resolve, reject) => {
    try {
      setTimeout(() => {
        resolve(req)
      }, 1000)
    } catch (err) {
      reject(err)
    }
  })

export const createUpdateGameRequest = game => ({
  method: 'PUT',
  mode: 'cors',
  headers: {
    'Content-Type': 'application/json',
  },
  body: JSON.stringify(game),
})

export const API_URL = process.env.API_URL || 'http://localhost:8080/'

export const createAction = (type, body) => ({ type, body })

const createRequest = (payload, method) => ({
  method,
  mode: 'cors',
  headers: {
    'Content-Type': 'application/json',
  },
  ...(method !== 'GET' && payload && { body: JSON.stringify(payload) }),
})

export const createThunk = ({
  action,
  url,
  payload,
  parse = ({ result }) => result,
  method,
}) => dispatch => {
  dispatch(createAction(`${action}_${BEGIN}`))
  return fetch(`${API_URL}${url}`, createRequest(payload, method))
    .then(result => result.json())
    .then(result => {
      dispatch(createAction(`${action}_${SUCCESS}`, parse({ result, type: SUCCESS })))
    })
    .catch(result => {
      dispatch(createAction(`${action}_${SUCCESS}`, parse({ result, type: ERROR })))
    })
}
