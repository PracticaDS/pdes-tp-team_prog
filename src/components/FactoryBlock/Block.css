.MachineNode {
  display: flex;
}

.Block {
  display: flex;
  margin: 0.1rem;
  border-radius: 10%;
  background: rgb(255, 255, 255);
  border: 0.1rem solid rgb(0, 0, 0);
  width: 3rem;
  height: 3rem;
}
