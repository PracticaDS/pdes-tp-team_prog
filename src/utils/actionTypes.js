export const RESTART_CURRENCY = 'restartCurrency'
export const SELECT_MACHINE = 'selectMachine'
export const UPDATE_BLOCK = 'updateBlock'
export const ROTATE_BLOCK = 'rotateBlock'
export const SELECT_ACTION = 'selectAction'
export const PLAY_GAME = 'playGame'
export const TICK = 'tick'
export const DELETE_BLOCK = 'deleteBlock'
export const CREATE_RAW_MATERIAL = 'CREATE_RAW_MATERIAL'
export const CREATE_ITEMS = 'CREATE_ITEMS'
export const DELETE_ITEMS = 'DELETE_ITEMS'
export const MOVE_BLOCK = 'moveSelectBlock'
export const SELECT_MOVE_BLOCK = 'selectMoveBlock'
export const DESELECT_MOVE_BLOCK = 'deselectMoveBlock'
export const BUY_MACHINE = 'buyMachine'
