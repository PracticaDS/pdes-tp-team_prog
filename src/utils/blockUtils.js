import { createEmptyMachine, createStarterMachine } from './machineUtils'

export const createEmptyBlock = ({ position, items = {} }) => ({
  position,
  machine: createEmptyMachine(),
  items,
})

export const createStarterBlock = ({ position, machine, items = {} }) => ({
  position,
  machine: machine || createStarterMachine({}),
  items,
})

// @Param block
// @Param item
export const updateBlockWithItem = ({ position, machine, items = {} }, item) => ({
  position,
  machine,
  items: {
    ...items,
    [item.name]: item,
  },
})
