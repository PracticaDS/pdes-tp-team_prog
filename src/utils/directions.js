/* eslint-disable no-restricted-syntax */
export const DOWN = 'Down'
export const UP = 'Up'
export const RIGHT = 'Right'
export const LEFT = 'Left'

export const applyDirection = (position, direction) => {
  if (direction === UP) return { ...position, row: position.row - 1 }
  if (direction === RIGHT) return { ...position, column: position.column + 1 }
  if (direction === DOWN) return { ...position, row: position.row + 1 }
  return { ...position, column: position.column - 1 }
}

export const isPositionValid = ({ row, column }, { n, m }) =>
  row < n && row >= 0 && column >= 0 && column < m
