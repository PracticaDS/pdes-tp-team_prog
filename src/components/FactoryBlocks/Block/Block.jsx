import React from 'react'
import './Block.css'

import { Card, CardContent } from '@material-ui/core'

// eslint-disable-next-line react/prop-types
const Block = ({ position: { row, column } }) => (
  <Card className="Block">
    <CardContent>
      {`Row: ${row}`}
      {'\n'}
      {`Column: ${column}`}
    </CardContent>
  </Card>
)

export default Block
