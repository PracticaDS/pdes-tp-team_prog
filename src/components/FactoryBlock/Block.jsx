import React from 'react'
import { Card } from '@material-ui/core'
import './Block.css'
import { getDegree } from '../../utils/directions'
import { displayAction } from './BlockUtils'

const Block = ({
  directionToRotate, // component
  isSelected, // component
  isEmptyBlock, // component
  renderNode, // component
  node, // store
  machineSelected,
  currency,
  moveSelectedNode, // store
  actionSelected, // store
  blockActions,
}) => (
  <div
    className={`block ${isSelected ? 'selected' : 'unselected'} ${
      !isEmptyBlock ? ` ${getDegree(directionToRotate)}` : ''
    }`}
    onClick={() =>
      displayAction({
        node,
        currency,
        machineSelected,
        moveSelectedNode,
        actionSelected,
        ...blockActions,
      })
    }
  >
    <Card>{renderNode(node)}</Card>
  </div>
)

export default Block
