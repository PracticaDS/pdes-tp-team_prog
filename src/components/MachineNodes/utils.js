import { TIMER_TIME } from '../../utils/defaultValues'
import { isPositionValid } from '../../utils/directions'

const TIMEOUT_TIME = TIMER_TIME / 2

export const incrementCount = prevState => ({
  personalCount: prevState.personalCount + 1,
})

export const resetCount = () => ({
  personalCount: 0,
})

export const deactivate = () => ({
  isActive: false,
})

export const activate = () => ({
  isActive: true,
})

export const activateNode = () => ({
  ...resetCount(),
  ...activate(),
})

export const withTimeout = cb => {
  setTimeout(() => {
    cb()
  }, TIMEOUT_TIME)
}

export const areAllMaterials = (necessaryMaterials, currentMaterials) =>
  necessaryMaterials.every(materialRecipe => {
    const materialInGrid = currentMaterials[materialRecipe.material.name]
    return materialInGrid && materialInGrid.quantity >= materialRecipe.quantity
  })

export const materialsFiltered = (necessaryMaterials, currentMaterials) =>
  necessaryMaterials.map(materialRecipe => {
    const materialInGrid = currentMaterials[materialRecipe.material.name]
    return { ...materialInGrid, quantity: materialRecipe.quantity }
  })

export const hasNecessaryMaterials = (machine, items) => {
  const { metadata } = machine
  const { recipeSelected } = metadata
  const { necessaryMaterials } = recipeSelected

  return areAllMaterials(necessaryMaterials, items)
}

export const getItemsFromMaterials = (machine, items) => {
  const { metadata } = machine
  const { recipeSelected } = metadata
  const { necessaryMaterials } = recipeSelected

  return materialsFiltered(necessaryMaterials, items)
}

export const getCraftedMaterial = machine => {
  const { metadata } = machine
  const { recipeSelected } = metadata
  const { result } = recipeSelected
  const { material, quantity } = result

  return { material, quantity }
}

// machine - state
export const isFrequencyFinished = ({ node }, { personalCount }) =>
  node.machine && node.machine.frequency && node.machine.frequency <= personalCount

export const runCBInValidPosition = (position, dimensions, cb) => {
  if (isPositionValid(position, dimensions)) {
    cb()
  }
}
