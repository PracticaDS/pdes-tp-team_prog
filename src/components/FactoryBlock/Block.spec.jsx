import React from 'react'
import { mount } from 'enzyme'
import Block from './Block'

import { DOWN, LEFT, RIGHT, UP, getDegree } from '../../utils/directions'

describe('Block Suite', () => {
  describe('Basic rendering', () => {
    let renderNodeMock
    let node

    beforeEach(() => {
      renderNodeMock = jest.fn()
      node = {
        position: {
          row: 0,
          column: 0,
        },
      }
    })

    it('render simple Block component', () => {
      const BlockComponent = mount(<Block node={node} renderNode={renderNodeMock} />)
      expect(BlockComponent).toBeDefined()
    })
  })
})

it('when calling getDegree with direction Left then returns css object with transform 90 degree', () => {
  const cssDown = getDegree(LEFT)
  expect(cssDown).toEqual('left')
})

it('when calling getDegree with direction Right then returns css object with transform 270 degree', () => {
  const cssDown = getDegree(RIGHT)
  expect(cssDown).toEqual('right')
})

it('when calling getDegree with direction Up then returns css object with transform 180 degree', () => {
  const cssDown = getDegree(UP)
  expect(cssDown).toEqual('up')
})

it('when calling getDegree with direction Down then returns css object with transform 0 degree', () => {
  const cssDown = getDegree(DOWN)
  expect(cssDown).toEqual('down')
})
