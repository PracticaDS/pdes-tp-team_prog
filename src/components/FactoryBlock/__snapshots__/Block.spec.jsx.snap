// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render simple Block component 1`] = `
<div
  className="Block"
  onClick={[Function]}
  style={
    Object {
      "transform": "rotate(0deg)",
    }
  }
>
  <div
    className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
  >
    <div
      className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
    >
      <img
        className="MuiCardMedia-root-29 MuiCardMedia-media-30"
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </div>
  </div>
</div>
`;
